uberfire-project/uberfire-project-backend/src/main/java/org/guvnor/common/services/project/backend/server/utils/configuration/ConfigurationStrategy.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2018 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.guvnor.common.services.project.backend.server.utils.configuration;
+
+import java.util.Map;
+
+/**
+ * Define the behaviour of a ConfigurationStrategy,
+ * load the configuration and check if it's valid
+ */
+public interface ConfigurationStrategy extends Valid,
+                                               Order {
+
+    Map<ConfigurationKey, String> loadConfiguration();
```

Comment:
This design of having separate `loadConfiguration` and `isValid` methods does not enforce correct usage of the API. For example user can load the configuration which is invalid and then forget to call `isValid()` and do something about invalid configurationd.

In my opinion ConfigurationStrategy should only have one method like 
`Map<ConfigurationKey, String> loadConfiguration() throws InvalidConfigurationException;` where `InvalidConfigurationException` would be checked exception, forcing the user to handle it.
Also that exception could hold details about what is invalid like: missing key? invalid/unparseable value of property? Invalid / unsupported combination of properties?

@desmax74 what do you think?

Reply:
The isValid is used by the ConfigurationContextProvider to understand if is ok the Strategy impl or if is better to use the next, using the exceptions to understand if is valid or not in this internal API isn't a good idea because a user will be forced to use a configuration not wanted to remove error messages.